web/api/v1/api.go

```
@@ -1740,6 +1742,40 @@ func TSDBStatsFromIndexStats(stats []index.Stat) []TSDBStat {
 	return result
 }
 
+func (api *API) serveTSDBBlock(_ *http.Request) apiFuncResult {
+	dirs, err := os.ReadDir(api.dbDir)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorInternal, fmt.Errorf("read TSDB dir: %w", err)}, nil, nil}
+	}
+
+	var blocks []tsdb.BlockMeta
+
+	for _, entry := range dirs {
+		if !entry.IsDir() {
+			continue
+		}
+		_, err := ulid.Parse(entry.Name())
+		if err != nil {
+			continue
+		}
+
+		blockPath := filepath.Join(api.dbDir, entry.Name())
+		block, err := tsdb.OpenBlock(api.logger, blockPath, nil, nil)
+		if err != nil {
+			continue
```

Is this `continue` here to handle deletions of underlying block directories while iterating over them? I think we should error out instead of just continuing unless it's an error that indicates that the directory doesn't exist anymore, right?

But this race condition, along with my other comment above about re-fetching the metadata for each block although we already have that information in the TSDB instance itself somewhere, make me wonder whether it would be better to add a function to the TSDB itself that can return this existing info. Then the TSDB instance could handle both the race condition / concurrency aspects (which I guess it already has to do anyway for other purposes), as well as handing out the existing info it has. I don't know too much about the current TSDB code though, does that make sense?